### PR TITLE
Improve finality manager performance and add more test simulations - Closes #3894

### LINKED ISSUE #3894: Improve finality manager performance and add more test simulations
We developed more data simulations for the voting, need to add to the tests. Also @limiaspasdaniel identified in his comments https://github.com/LiskHQ/lisk-sdk/pull/3851#discussion_r298516975 https://github.com/LiskHQ/lisk-sdk/pull/3851#discussion_r298517614 some points that can improve the performance. 

### Which version(s) does this affect? (Environment, OS, etc...)
3.0.0

### PR DESCRIPTION
**Pull request by nazarhussain**
### What was the problem?

We created more data simulations for the BFT, that was required to be added. 

### How did I solve it?

Added the CSV files and improved the test code. 

### How to manually test it?

```
nom run jest:unit finality_manager
```

### Review checklist

- [x] The PR resolves #3894
- [ ] All new code is covered with unit tests
- [ ] Relevant integration / functional tests are added
- [ ] Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
- [ ] Documentation has been added/updated

### CODE REVIEW

File: framework/src/modules/chain/bft/finality_manager.js [hunk 1]
```diff
@@ -200,20 +204,20 @@ class FinalityManager {
...
-		const higherPairVoted = Object.entries(this.preVotes)
+		const highestHeightPreVoted = Object.keys(this.preVotes)
 			.reverse()
-			.find(pair => pair[1] >= this.preVoteThreshold);
+			.find(key => this.preVotes[key] >= this.preVoteThreshold);
 
-		this.prevotedConfirmedHeight = higherPairVoted
-			? parseInt(higherPairVoted[0], 10)
+		this.prevotedConfirmedHeight = highestHeightPreVoted
+			? parseInt(highestHeightPreVoted, 10)
```
Comment (limiaspasdaniel) [inline framework/src/modules/chain/bft/finality_manager.js]: Do we need to specify the base?

File: framework/test/jest/unit/specs/modules/chain/bft/bft_manager_scenarios.spec.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+	describe('addBlockHeader', () => {
+		BFTScenarios.forEach(scenario => {
+			describe(`when ${scenario.title}`, () => {
+				const myBft = new FinalityManager({
+					finalizedHeight: scenario.config.finalizedHeight,
+					activeDelegates: scenario.config.activeDelegates,
+				});
+
+				scenario.steps.forEach(step => {
+					it(`have accurate information when ${
```
Comment (limiaspasdaniel) [inline framework/test/jest/unit/specs/modules/chain/bft/bft_manager_scenarios.spec.js]:
```suggestion
					it(`should have accurate information when ${
```

File: framework/src/modules/chain/bft/finality_manager.js (hunk 1 above)
Comment (nazarhussain) [reply framework/src/modules/chain/bft/finality_manager.js]: Some es-lint rule introduced, which required to specify it.

**Comment by nazarhussain** [comment]
@shuse2 @limiaspasdaniel @michielmulders I found some anomaly in the dataset, will figure it out and ask for re-review.

File: framework/src/modules/chain/bft/finality_manager.js [hunk 3]
```diff
@@ -200,39 +209,85 @@ class FinalityManager {
...
+	 * @return {number}
+	 * @private
+	 */
+	_getValidMinHeightToCommit(header) {
+		const searchTillHeight = Math.max(
+			this.minHeight,
+			header.height - this.processingThreshold,
+		);
+
+		let needleHeight = Math.max(
```
Comment (michielmulders) [inline framework/src/modules/chain/bft/finality_manager.js]: Just curious why you choose for the variable name `needleHeight`?

Comment (michielmulders) [reply framework/src/modules/chain/bft/finality_manager.js]: Does it refer to latest writing height?

File: framework/src/modules/chain/bft/finality_manager.js [hunk 4]
```diff
@@ -290,6 +345,26 @@ class FinalityManager {
 		return true;
 	}
 
+	/**
+	 * Cleanup pre-votes and pre-commits objects
+	 * to only keep the track of maximum 5 rounds (headers list size)
+	 *
+	 * @private
+	 */
+	_cleanup() {
```
Comment (michielmulders) [inline framework/src/modules/chain/bft/finality_manager.js]: Is this slicing only one round or 505 headers?
Worth mentioning in the description!

File: framework/test/jest/unit/specs/modules/chain/bft/bft_manager_scenarios.spec.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+const BFTScenarios = require('./scenarios');
+
+const {
+	FinalityManager,
+} = require('../../../../../../../src/modules/chain/bft/finality_manager');
+
+describe('FinalityManager', () => {
+	describe('addBlockHeader', () => {
+		BFTScenarios.forEach(scenario => {
```
Comment (michielmulders) [inline framework/test/jest/unit/specs/modules/chain/bft/bft_manager_scenarios.spec.js]: Very clean, nice!

File: framework/src/modules/chain/bft/finality_manager.js (hunk 3 above)
Comment (nazarhussain) [reply framework/src/modules/chain/bft/finality_manager.js]: Its an old term we learned in school while iterating over the complex structure. That you put a needle as a cursor.

File: framework/src/modules/chain/bft/finality_manager.js (hunk 4 above)
Comment (nazarhussain) [reply framework/src/modules/chain/bft/finality_manager.js]: 505 headers, I will add few comments further.

Comment (nazarhussain) [reply framework/src/modules/chain/bft/finality_manager.js]: Its mentioned in the function description in comments that its cleaning 5 rounds.

**Pull request merged**